url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/quartz/triggers/CronExpression.java

```diff
@@ -0,0 +1,1571 @@
+    // the next day, month etc. This is done by adding the maximum amount for that
+    // type, and using modulus max to determine the value being added.
+    int max = -1;
+    if (stopAt < startAt) {
+      switch (type) {
+      case SECOND:
+        max = 60;
+        break;
+      case MINUTE:
+        max = 60;
+        break;
+      case HOUR:
+        max = 24;
+        break;
+      case MONTH:
+        max = 12;
+        break;
+      case DAY_OF_WEEK:
+        max = 7;
+        break;
+      case DAY_OF_MONTH:
+        max = 31;
+        break;
+      case YEAR:
+        throw new IllegalArgumentException("Start year must be less than stop year");
+      default:
+        throw new IllegalArgumentException("Unexpected type encountered");
+      }
+      stopAt += max;
+    }
+
+    for (int i = startAt; i <= stopAt; i += incr) {
+      if (max == -1) {
+        // ie: there's no max to overflow over
+        set.add(i);
+      } else {
+        // take the modulus to get the real value
+        int i2 = i % max;
+
+        // 1-indexed ranges should not include 0, and should include their max
+        if (i2 == 0 && (type == MONTH || type == DAY_OF_WEEK || type == DAY_OF_MONTH)) {
+          i2 = max;
+        }
+
+        set.add(i2);
+      }
+    }
+  }
+
+  TreeSet<Integer> getSet(int type) {
+    switch (type) {
+    case SECOND:
+      return seconds;
+    case MINUTE:
+      return minutes;
+    case HOUR:
+      return hours;
+    case DAY_OF_MONTH:
+      return daysOfMonth;
+    case MONTH:
+      return months;
+    case DAY_OF_WEEK:
+      return daysOfWeek;
+    case YEAR:
+      return years;
+    default:
+      return null;
+    }
+  }
+
+  protected ValueSet getValue(int v, String s, int i) {
+    char c = s.charAt(i);
+    StringBuilder s1 = new StringBuilder(String.valueOf(v));
+    while (c >= '0' && c <= '9') {
+      s1.append(c);
+      i++;
+      if (i >= s.length()) {
+        break;
+      }
+      c = s.charAt(i);
+    }
+    ValueSet val = new ValueSet();
+
+    val.pos = (i < s.length()) ? i : i + 1;
+    val.value = Integer.parseInt(s1.toString());
+    return val;
+  }
+
+  protected int getNumericValue(String s, int i) {
+    int endOfVal = findNextWhiteSpace(i, s);
+    String val = s.substring(i, endOfVal);
+    return Integer.parseInt(val);
+  }
+
+  protected int getMonthNumber(String s) {
+    Integer integer = monthMap.get(s);
+
+    if (integer == null) {
+      return -1;
+    }
+
+    return integer;
+  }
+
+  protected int getDayOfWeekNumber(String s) {
+    Integer integer = dayMap.get(s);
+
+    if (integer == null) {
+      return -1;
+    }
+
+    return integer;
+  }
+
+  ////////////////////////////////////////////////////////////////////////////
+  //
+  // Computation Functions
+  //
+  ////////////////////////////////////////////////////////////////////////////
+
+  public Date getTimeAfter(Date afterTime) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 150 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.